native/spark-expr/src/hash_funcs/utils.rs

```diff
@@ -106,17 +106,28 @@ macro_rules! hash_array_primitive_float {
 
 #[macro_export]
 macro_rules! hash_array_decimal {
-    ($array_type:ident, $column: ident, $hashes: ident, $hash_method: ident) => {
+    ($array_type:ident, $column: ident, $hashes: ident, $hash_method: ident, $precision: ident) => {
         let array = $column.as_any().downcast_ref::<$array_type>().unwrap();
+        let precision = *$precision;
 
         if array.null_count() == 0 {
             for (i, hash) in $hashes.iter_mut().enumerate() {
-                *hash = $hash_method(array.value(i).to_le_bytes(), *hash);
+                // when precision is less than or equal to 18, spark use deimal long value to calculate hash.
+                // refer: org.apache.spark.sql.catalyst.expressions.InterpretedHashFunction.hash
+                if precision <= 18 {
```

Comment:
I wonder if we should create a separate version of the macro for the `precision <= 18` case to avoid branching?

Reply:
Thank you, that sounds good to me, I tried separating it